lib/posix/pthread.c

```diff
@@ -135,17 +153,44 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
 	uint32_t pthread_num;
 	pthread_condattr_t cond_attr;
 	struct posix_thread *thread;
+	pthread_attr_t dynamic_attr;
+	k_thread_stack_t *dynamic_stack = NULL;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (attr != NULL && (attr->initialized == 0 || (attr->stack == NULL
+		&& attr->stacksize < ARCH_MIN_STACK_SIZE))) {
 		return EINVAL;
 	}
 
+	if (attr == NULL || attr->stack == NULL) {
+		if (IS_ENABLED(CONFIG_PTHREAD_DYNAMIC_STACK)) {
+			/*
+			 * We dynamically allocate space when either
+			 * 1) attr == NULL -> use DYNAMIC_STACK_SIZE, or
+			 * 2) attr != NULL && attr->stack == NULL
+			 *    -> allocate attr->stacksize
+			 */
+			if (attr == NULL) {
+				(void) pthread_attr_init(&dynamic_attr);
+				dynamic_attr.stacksize = DYNAMIC_STACK_SIZE;
+				attr = &dynamic_attr;
+			}
+
+			dynamic_stack = k_malloc(
+				Z_KERNEL_STACK_SIZE_ADJUST(attr->stacksize));
+			if (dynamic_stack == NULL) {
+				return EAGAIN;
+			}
+
+			*((void **) &attr->stack) = dynamic_stack;
+			*((uint8_t **) &attr->stack) =
+				(uint8_t *)ROUND_UP(attr->stack,
+				Z_KERNEL_STACK_OBJ_ALIGN);
+			*((uint32_t *) &attr->flags) |= K_STACK_ON_HEAP;
```

Comment:
Does it also make sense to or with K_INHERIT_PERMS in this case? cc: @andrewboie

Reply:
all this casting here is making this difficult to parse. is there any way we can adjust types so all this isn't necessary? 
also this doesn't look right that you are rounding the pointer up here, you need to use an allocator that supports alignment. k_heap has a facility for aligned allocations you should be using that instead of k_malloc to reserve memory

Reply:
Agreed - it is difficult to parse. The argument to `pthread_create(3)` is const. I will add another local variable that shadows it instead.

`k_heap_alloc()` doesn't seem to accept any alignment arguments - will that be sufficient for thread stack alignment?

Reply:
I've opted to use `posix_memalign(3)` here because that seems to be the most appropriate solution for posix thread, and then I'll worry about the implementation of that in #29519.

Reply:
Wow - `k_mem_pool`, `sys_mem_pool`, and `k_heap`. This is really confusing. I see what you mean now that the  `k_heap` API does handle alignment with `k_heap_aligned_alloc()`. Just trying to figure out how to unbury that is quite a challenge.